packages/react/src/core/collection/index.tsx

```diff
@@ -0,0 +1,789 @@
+  args: CollectionUpdateApiArgs<TContext, TFields> & {
+    defaultApi: CollectionUpdateApiHandler<TContext, TFields>
+  }
+) => Promisable<CollectionUpdateApiReturn>
+
+export type CollectionUpdateDefaultApiArgs<
+  TContext extends AnyContextable,
+  TFields extends Fields,
+> = ServerApiHandlerArgs<TContext, TFields> & {
+  // This should be the primary field of the collection e.g. __pk or username
+  id: string | number
+}
+
+// TODO: This is not correct, it should return default value of form instead of just simple findOne response
+export type CollectionUpdateDefaultApiReturn<TFields extends Fields> = any
+
+export type CollectionUpdateDefaultApiHandler<
+  TContext extends AnyContextable,
+  TFields extends Fields,
+> = (
+  args: CollectionUpdateDefaultApiArgs<TContext, TFields> & {
+    defaultApi: CollectionUpdateDefaultApiHandler<TContext, TFields>
+  }
+) => Promisable<CollectionUpdateDefaultApiReturn<TFields>>
+
+export type CollectionUpdateConfig<
+  TContext extends AnyContextable = AnyContextable,
+  TFields extends Fields = Fields,
+> = {
+  updateApi?: CollectionUpdateApiHandler<TContext, TFields>
+  updateDefaultApi?: CollectionUpdateDefaultApiHandler<TContext, TFields>
+  options?: Simplify<FieldsOptions<TContext, TFields>>
+}
+
+export interface CollectionDeleteApiArgs<TContext extends AnyContextable, TFields extends Fields>
+  extends ServerApiHandlerArgs<TContext, TFields> {
+  ids: string[] | number[]
+}
+
+export type CollectionDeleteApiReturn = {
+  success: boolean
+}
+
+export type CollectionDeleteApiHandler<TContext extends AnyContextable, TFields extends Fields> = (
+  args: CollectionDeleteApiArgs<TContext, TFields> & {
+    defaultApi: CollectionDeleteApiHandler<TContext, TFields>
+  }
+) => Promisable<CollectionDeleteApiReturn>
+
+// Extract searchable columns from fields
+export type ExtractSearchableColumns<TFields extends Fields> = {
+  [K in keyof TFields['shape']]: TFields['shape'][K] extends FieldColumnShape
+    ? TFields['shape'][K]['$client']['column']['dataType'] extends typeof DataType.STRING
+      ? Extract<K, string>
+      : never
+    : never
+}[keyof TFields['shape']]
+
+// Extract sortable columns from fields
+export type ExtractSortableColumns<TFields extends Fields> = {
+  [K in keyof TFields['shape']]: TFields['shape'][K] extends FieldColumnShape
+    ? TFields['shape'][K]['$client']['column']['dataType'] extends
+        | typeof DataType.STRING
+        | typeof DataType.INT
+        | typeof DataType.FLOAT
+        | typeof DataType.DATETIME
+        | typeof DataType.BIGINT
+        | typeof DataType.DECIMAL
+      ? Extract<K, string>
+      : never
+    : never
+}[keyof TFields['shape']]
+
+// ListConfiguration
+export interface ListConfiguration<TFields extends Fields> {
+  search?: ExtractSearchableColumns<TFields>[]
+  sortBy?: ExtractSortableColumns<TFields>[]
+}
+
+export interface CollectionListResponse {
+  data: ({ __id: string | number; __pk: string | number } & {})[]
```

Comment:
Why do we need to `& {}` before using ?

Reply:
This would be `({ __id: string | number; __pk: string | number } & Record<string, unknown>)[]`